File: src/chapters/nonlinear-equations/content/code/broyden.r
broyden <- function(F, x0, tol = 1e-12, max_iter = 100) {
  # Broyden's (good) method for F(x) = 0; approximates the Jacobian.
  x <- as.numeric(x0)
  n <- length(x)
  B <- diag(n)                          # initial Jacobian approximation
  Fx <- as.numeric(F(x))
  for (k in 1:max_iter) {
    if (max(abs(Fx)) < tol) return(list(root = x, iters = k))
    dx <- as.numeric(solve(B, -Fx))
    x <- x + dx
    Fx_new <- as.numeric(F(x))
    dF <- Fx_new - Fx
    B <- B + outer(dF - as.numeric(B %*% dx), dx) / sum(dx * dx)   # rank-1 update
    Fx <- Fx_new
  }
  list(root = x, iters = max_iter)
}

F <- function(v) c(v[1]^2 + v[2]^2 - 4, v[1] * v[2] - 1)
res <- broyden(F, c(2.0, 0.5))
cat(sprintf("(array([%.8f, %.8f]), %d)\n", res$root[1], res$root[2], res$iters))

File: src/chapters/nonlinear-equations/content/code/broyden.jl
using LinearAlgebra

function broyden(F, x0; tol = 1e-12, max_iter = 100)
    x = float.(x0); n = length(x); B = Matrix{Float64}(I, n, n); Fx = F(x)
    for k in 1:max_iter
        maximum(abs.(Fx)) < tol && return x, k
        dx = B \ (-Fx); x = x + dx
        Fn = F(x); y = Fn - Fx
        B = B + ((y - B*dx) * dx') / (dx' * dx)     # rank-1 update
        Fx = Fn
    end
    return x, max_iter
end
F(v) = [v[1]^2 + v[2]^2 - 4, v[1]*v[2] - 1]
println(broyden(F, [2.0, 0.5]))
